.github/workflows/webapp-shared.yml

```diff
@@ -0,0 +1,53 @@
+name: Shared Web App
+
+on:
+  workflow_call:
+    inputs:
+      workspace-argument:
+        required: true
+        type: string
+
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  ci:
+    runs-on: ubuntu-22.04
+    defaults:
+      run:
+        working-directory: webapp
+    steps:
+      - name: ci/checkout-repo
+        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+      - name: ci/setup-node
+        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # v3.6.0
+        id: setup_node
+        with:
+          node-version-file: ".nvmrc"
+          cache: npm
+          cache-dependency-path: 'webapp/package-lock.json'
+      - name: ci/get-node-modules
+        run: |
+          make node_modules
+      - name: ci/check-lint
+        id: check-lint
+        run: |
+          npm run check ${{ inputs.workspace-argument }}
+        continue-on-error: true
+      - name: ci/check-types
+        id: check-types
+        run: |
+          npm run check-types ${{ inputs.workspace-argument }}
+        continue-on-error: true
+      - name: ci/test
+        id: test
+        run: |
+          npm test ${{ inputs.workspace-argument }}
+        continue-on-error: true
+      - name: ci/check-for-failures
+        if: steps.check-lint.outcome != 'success' || steps.check-types.outcome != 'success' || steps.test.outcome != 'success'
```

Comment:
If any of the `check-lint. outcome `,`check-types` & `test.outcome` steps will fail by returning non zero exit code, what is the reason for introducing this conditional, along with `continue-on-error` on all the check steps, and the forced `exit 1` ?

Reply:
I'm going to see if I can get the output of `ci/check-for-failures` to give a bit more information on that as well since it's not clear to someone seeing that that they'd need to look at the other steps to see what actually failed

Reply:
I added some extra info here about what failed so that people aren't completely lost if they see that
![Screen Shot 2023-04-25 at 11 20 06 AM](https://user-images.githubusercontent.com/3277310/234324553-110687e9-5162-472b-b153-652b8f9e4f02.png)

Reply:
Hmm, What about using different steps for each Test? 
In that way, the tests will be parallelised, all checks will run, and correct output will be presented to the user.

Reply:
This will also provide a better, more concise, simpler workflow 

Removing conditionals, and the `ci/check-for-failures` step.

Reply:
So essentially the problem is there is no way to say mark as failure but continue? Could we just make them separate jobs?

Reply:
Didn't see @phoinixgrr 's reply. Probably what he said. I really don't want to have to hunt for my failures.

Reply:
Separate jobs would mean that we're back with what we had before where there's separate checks for linting, tests, etc flooding the list of checks at the end of the PR which is another thing this PR is trying to improve.

The messages I added to that step will tell people where to go for the actual failure, so it's only 1 extra click to see that information than it was before

Reply:
I am referring to this: https://docs.github.com/en/actions/using-jobs/using-jobs-in-a-workflow

As an end user I would find more intuitive getting a quick info that linting or check-types testing has failed for my changes being introduced, instead of going through a pipeline to read the custom failure written. 

I would strongly recommend separating the tests into jobs, and creating a seperate job for the build step, that has dependency on all of the tests, for the following reasons: 
- Better user experience 
- Faster execution with jobs inherently being parallelised
- Better readability for the CI and simpler workflow (removing conditionals and extra custom steps) 

But feel free to merge this as is, if it suits your needs better.

Reply:
One of the goals of this was specifically getting away from that actually since every job corresponds to one check in the list at the bottom of each PR. That said, I'm realizing the separate jobs per web app package wouldn't scale well.

Between that and the concerns about making CI too granular, we've decided to scale back the changes we're going to make here. I'll submit a followup tomorrow where we're just cleaning up the pipeline a bit and adding the path filtering for the web app, but otherwise, the structure will be more or less the same as it was before.